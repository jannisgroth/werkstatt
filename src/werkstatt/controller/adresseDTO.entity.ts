/* eslint-disable @eslint-community/eslint-comments/disable-enable-pair */
/* eslint-disable @typescript-eslint/no-magic-numbers */

/**
 * Das Modul besteht aus der Entity-Klasse.
 * @packageDocumentation
 */
import { ApiProperty } from '@nestjs/swagger';
import { Matches, MaxLength } from 'class-validator';

export class AdresseDTO {
    @Matches(/^\d{5}$/u)
    @MaxLength(5)
    @ApiProperty({ example: '76227', type: String })
    readonly postleitzahl!: string;

    @Matches(/^[\p{L}\s\-']+$/iu)
    @MaxLength(30)
    @ApiProperty({ example: 'Mozartstra√üe', type: String })
    readonly strasse: string | undefined;

    @Matches(/^\d{1,2}[A-Za-z]?$/u)
    @MaxLength(4)
    @ApiProperty({ example: '12A', type: String })
    readonly hausnummer: string | undefined;
}
