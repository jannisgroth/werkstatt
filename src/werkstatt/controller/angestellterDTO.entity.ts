// eslint-disable-next-line @eslint-community/eslint-comments/disable-enable-pair
/* eslint-disable @typescript-eslint/no-magic-numbers */

/**
 * Das Modul besteht aus der Entity Klasse.
 * @packageDocumentation
 */
import { ApiProperty } from '@nestjs/swagger';
import { IsOptional, Matches, MaxLength } from 'class-validator';

/**
 * Entity-Klasse für Angestelle ohne TypeORM.
 */
export class AngestellterDTO {
    @Matches(/^[\sa-zßäöü-]+$/iu)
    @MaxLength(30)
    @ApiProperty({ example: 'Der Name', type: String })
    readonly name!: string;

    @IsOptional()
    @MaxLength(20)
    @ApiProperty({ example: 'Projektleiter', type: String })
    readonly position: string | undefined;
}
