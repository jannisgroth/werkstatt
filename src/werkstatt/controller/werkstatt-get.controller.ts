// eslint-disable-next-line max-classes-per-file
import {
    Controller,
    Get,
    Headers,
    HttpStatus,
    NotFoundException,
    Param,
    Query,
    Req,
    Res,
    UseInterceptors,
} from '@nestjs/common';
import {
    ApiHeader,
    ApiNotFoundResponse,
    ApiOkResponse,
    ApiOperation,
    ApiParam,
    ApiProperty,
    ApiResponse,
    ApiTags,
} from '@nestjs/swagger';
import { Request, Response } from 'express';
import { Public } from 'nest-keycloak-connect';
import { paths } from '../../config/paths.js';
import { getLogger } from '../../logger/logger.js';
import { ResponseTimeInterceptor } from '../../logger/response-time.interceptor.js';
import { Adresse } from '../entity/adresse.entity.js';
import { Werkstatt, WerkstattArt } from '../entity/werkstatt.entity.js';
import { Suchkriterien } from '../service/suchkriterien.js';
import { WerkstattReadService } from '../service/werkstatt-read.service.js';
import { getBaseUri } from './getBaseUri.js';

export type Link = {
    readonly href: string;
};

export type Links = {
    /** self-Link */
    readonly self: Link;
    /** Optionaler Linke für list */
    readonly list?: Link;
    /** Optionaler Linke für add */
    readonly add?: Link;
    /** Optionaler Linke für update */
    readonly update?: Link;
    /** Optionaler Linke für remove */
    readonly remove?: Link;
};

export type AdresseModel = Omit<Adresse, 'werkstatt' | 'id'>;

export type WerkstattModel = Omit<
    Werkstatt,
    'angestellte' | 'aktualisiert' | 'erzeugt' | 'id' | 'adresse' | 'version'
> & {
    adresse: AdresseModel;

    // eslint-disable-next-line @typescript-eslint/naming-convention
    _links: Links;
};

export type WerkstaetteModel = {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    _embedded: {
        werkstaette: WerkstattModel[];
    };
};

export class WerksatattQuery implements Suchkriterien {
    @ApiProperty({ required: false })
    declare readonly name: string;

    @ApiProperty({ required: false })
    declare readonly rating: number;

    @ApiProperty({ required: false })
    declare readonly art: WerkstattArt;

    @ApiProperty({ required: false })
    declare readonly homepage: string;

    @ApiProperty({ required: false })
    declare readonly fertigbau: string;

    @ApiProperty({ required: false })
    declare readonly rohbau: string;

    @ApiProperty({ required: false })
    declare readonly adresse: string;
}

const APPLICATION_HAL_JSON = 'application/hal+json';

@Controller(paths.rest)
@UseInterceptors(ResponseTimeInterceptor)
@ApiTags('Werkstatt REST-API')
export class WerkstattGetController {
    readonly #service: WerkstattReadService;

    readonly #logger = getLogger(WerkstattGetController.name);

    constructor(service: WerkstattReadService) {
        this.#service = service;
    }

    // eslint-disable-next-line max-params
    @Get(':id')
    @Public()
    @ApiOperation({ summary: 'Sucher mit der Werkstatt-ID' })
    @ApiParam({
        name: 'id',
        description: 'Z.B. 1',
    })
    @ApiHeader({
        name: 'If-None-Match',
        description: 'Header für bedingte Get-Requests, zb "0"',
        required: false,
    })
    @ApiOkResponse({ description: 'Die Werkstatt wurde gefunden' })
    @ApiNotFoundResponse({ description: 'Keine Werkstatt zur ID gefunden' })
    @ApiResponse({
        status: HttpStatus.NOT_MODIFIED,
        description: 'Die Werkstatt wurde bereits heruntergeladen',
    })
    async getById(
        @Param('id') idStr: string,
        @Req() req: Request,
        @Headers('If-None-Match') version: string | undefined,
        @Res() res: Response,
    ): Promise<Response<WerkstattModel | undefined>> {
        this.#logger.debug('getById: idStr=%s, version=%s', idStr, version);
        const id = Number(idStr);
        if (!Number.isInteger(id)) {
            this.#logger.debug('getById: not isInteger()');
            throw new NotFoundException(
                `Die Werkstatt-ID ${id} ist nicht gültig`,
            );
        }

        if (req.accepts([APPLICATION_HAL_JSON, 'json', 'html']) === false) {
            this.#logger.debug('getById: accepted=%o', req.accepted);
            return res.sendStatus(HttpStatus.NOT_ACCEPTABLE);
        }

        const werkstatt = await this.#service.findById({ id });
        if (this.#logger.isLevelEnabled('debug')) {
            this.#logger.debug('getById(): werkstatt=%s', werkstatt.toString());
            this.#logger.debug('getById(): adresse=%o', werkstatt.adresse);
        }

        const versionDb = werkstatt.version;
        if (version === `"${versionDb}"`) {
            this.#logger.debug('getById: NOT_MODIFIED');
            return res.sendStatus(HttpStatus.NOT_MODIFIED);
        }

        this.#logger.debug('getById: versionDb=%s', versionDb);
        res.header('ETag', `"${versionDb}"`);

        const werkstattModel = this.#toModel(werkstatt, req);
        this.#logger.debug('getById: werkstattModel=%o', werkstattModel);
        return res.contentType(APPLICATION_HAL_JSON).json(werkstattModel);
    }

    @Get()
    @Public()
    @ApiOperation({ summary: 'Suche mit Suchkriteren' })
    @ApiResponse({ description: 'Eine eventuel leere Liste ' })
    async get(
        @Query() query: WerksatattQuery,
        @Req() req: Request,
        @Res() res: Response,
    ): Promise<Response<WerkstattModel | undefined>> {
        this.#logger.debug('get: query%o', query);

        if (req.accepts([APPLICATION_HAL_JSON, 'json', 'html']) === false) {
            this.#logger.debug('get accepted=%o', req.accepted);
            return res.sendStatus(HttpStatus.NOT_ACCEPTABLE);
        }

        const werkstaette = await this.#service.find(query);
        this.#logger.debug('get %o', werkstaette);

        const werkstaetteModel: WerkstattModel[] = werkstaette.map(
            (werkstatt): WerkstattModel => this.#toModel(werkstatt, req, false),
        );
        this.#logger.debug('get werkstaetteModel=%o', werkstaetteModel);

        const result: WerkstaetteModel = {
            _embedded: { werkstaette: werkstaetteModel },
        };
        return res.contentType(APPLICATION_HAL_JSON).json(result).send();
    }

    #toModel(werkstatt: Werkstatt, req: Request, all = true) {
        const baseUri = getBaseUri(req);
        this.#logger.debug('#toModel: baseUri=%s', baseUri);
        const { id } = werkstatt;
        const links = all
            ? {
                  self: { href: `${baseUri}/${id}` },
                  list: { href: `${baseUri}` },
                  add: { href: `${baseUri}` },
                  update: { href: `${baseUri}/${id}` },
                  remove: { href: `${baseUri}/${id}` },
              }
            : { self: { href: `${baseUri}/${id}` } };

        this.#logger.debug(
            '#toModel: werkstatt=%o, links=%o',
            werkstatt,
            links,
        );
        const adresseModel: AdresseModel = {
            postleitzahl: werkstatt.adresse?.postleitzahl ?? 'N/A',
            strasse: werkstatt.adresse?.strasse ?? 'N/A',
            hausnummer: werkstatt.adresse?.hausnummer ?? 'N/A',
        };
        const werkstattModel: WerkstattModel = {
            werkstattname: werkstatt.werkstattname,
            rating: werkstatt.rating,
            art: werkstatt.art,
            homepage: werkstatt.homepage,
            schlagwoerter: werkstatt.schlagwoerter,
            adresse: adresseModel,
            _links: links,
        };

        return werkstattModel;
    }
}
