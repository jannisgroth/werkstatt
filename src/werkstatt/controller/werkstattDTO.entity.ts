/* eslint-disable @eslint-community/eslint-comments/disable-enable-pair */
/* eslint-disable max-classes-per-file */

/**
 * Das Modul besteht aus der Entity-Klasse.
 * @packageDocumentation
 */
import { ApiProperty } from '@nestjs/swagger';
import { Type } from 'class-transformer';
import {
    ArrayUnique,
    IsArray,
    IsInt,
    IsOptional,
    IsUrl,
    Matches,
    Max,
    Min,
    ValidateNested,
} from 'class-validator';
import { WerkstattArt } from '../entity/werkstatt.entity.js';
import { AdresseDTO } from './adresseDTO.entity.js';
import { AngestellterDTO } from './angestellterDTO.entity.js';

export const MAX_RATING = 5;

/**
 * Entity-Klasse für eine Werkstatt ohne TypeORM und ohne Referenzen.
 */
export class WerkstattDtoOhneRef {
    @ApiProperty({ example: 'Schreinerei Müller', type: String })
    readonly werkstattname!: string;

    @IsInt()
    @Min(0)
    @Max(MAX_RATING)
    @IsOptional()
    @ApiProperty({ example: 4, type: Number })
    readonly rating: number | undefined;

    @Matches(/^(SCHREINEREI|MALEREI|ELEKTRONIKER)$/u)
    @IsOptional()
    @ApiProperty({ example: 'Schreinerei', type: String })
    readonly art: WerkstattArt | undefined;

    @IsUrl()
    @IsOptional()
    @ApiProperty({ example: 'https://Etest.de/', type: String })
    readonly homepage: string | undefined;

    @IsOptional()
    @ArrayUnique()
    @ApiProperty({ example: ['FERTIGBAU', 'ROHBAU', 'INSTALLATION'] })
    readonly schlagwoerter: string[] | undefined;
}

/**
 * Entity-Klasse für eine Werkstatt ohne TypeORM.
 */
export class WerkstattDTO extends WerkstattDtoOhneRef {
    @ValidateNested()
    @Type(() => AdresseDTO)
    @ApiProperty({ type: AdresseDTO })
    readonly adresse!: AdresseDTO;

    @IsOptional()
    @IsArray()
    @ValidateNested()
    @Type(() => AngestellterDTO)
    @ApiProperty({ type: AngestellterDTO })
    readonly angestellte: AngestellterDTO[] | undefined;
}
