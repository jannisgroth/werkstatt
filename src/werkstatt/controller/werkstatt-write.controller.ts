import {
    Body,
    Controller,
    Delete,
    Headers,
    HttpCode,
    HttpStatus,
    Param,
    Post,
    Put,
    Req,
    Res,
    UseGuards,
    UseInterceptors,
} from '@nestjs/common';
import {
    ApiBadRequestResponse,
    ApiBearerAuth,
    ApiCreatedResponse,
    ApiForbiddenResponse,
    ApiHeader,
    ApiNoContentResponse,
    ApiOperation,
    ApiPreconditionFailedResponse,
    ApiResponse,
    ApiTags,
} from '@nestjs/swagger';
import { Request, Response } from 'express';
import { AuthGuard, Roles } from 'nest-keycloak-connect';
import { paths } from '../../config/paths.js';
import { getLogger } from '../../logger/logger.js';
import { ResponseTimeInterceptor } from '../../logger/response-time.interceptor.js';
import { type Adresse } from '../entity/adresse.entity.js';
import { type Angestellter } from '../entity/angestellter.entity.js';
import { type Werkstatt } from '../entity/werkstatt.entity.js';
import { WerkstattWriteService } from '../service/werkstatt-write.service.js';
import { getBaseUri } from './getBaseUri.js';
import { WerkstattGetController } from './werkstatt-get.controller.js';
import { WerkstattDTO, WerkstattDtoOhneRef } from './werkstattDTO.entity.js';

const MSG_FORBIDDEN = 'Kein Token mit ausreichender Berechtigung vorhanden';

@Controller(paths.rest)
@UseGuards(AuthGuard)
@UseInterceptors(ResponseTimeInterceptor)
@ApiTags('Werkstatt REST_API')
@ApiBearerAuth()
export class WerkstattWriteController {
    readonly #service: WerkstattWriteService;
    readonly #logger = getLogger(WerkstattGetController.name);

    constructor(service: WerkstattWriteService) {
        this.#service = service;
    }

    @Post()
    @Roles({ roles: ['admin', 'user'] })
    @ApiOperation({ summary: 'Eine neue Werkstatt anlegen' })
    @ApiCreatedResponse({ description: 'Erfolgreich neu angelegt' })
    @ApiBadRequestResponse({ description: 'Fehlerhafte Werkstattdaten' })
    @ApiForbiddenResponse({ description: MSG_FORBIDDEN })
    async post(
        @Body() werkstattDTO: WerkstattDTO,
        @Req() req: Request,
        @Res() res: Response,
    ): Promise<Response> {
        this.#logger.debug('post: werkstattDTO=%o', werkstattDTO);

        const werkstatt = this.#werkstattDtoToWerkstatt(werkstattDTO);
        const id = await this.#service.create(werkstatt);

        const location = `${getBaseUri(req)}/${id}`;
        this.#logger.debug('post: location=%s', location);
        return res.location(location).send();
    }

    // eslint-disable-next-line max-params
    @Put(':id')
    @Roles({ roles: ['admin', 'user'] })
    @HttpCode(HttpStatus.NO_CONTENT)
    @ApiOperation({
        summary: 'Eine vorhandene Werkstatt aktualisieren',
        tags: ['Aktualisieren'],
    })
    @ApiHeader({
        name: 'If-Match',
        description: 'Header für die optimistische Synchronisierung',
        required: false,
    })
    @ApiNoContentResponse({ description: 'Erfolgreich aktualisiert' })
    @ApiBadRequestResponse({ description: 'Fehlerhafte Werkstattdaten' })
    @ApiPreconditionFailedResponse({
        description: 'Falsche Version im Header "If-Match"',
    })
    @ApiResponse({
        status: HttpStatus.PRECONDITION_REQUIRED,
        description: 'Header "If-Match" fehlt',
    })
    @ApiForbiddenResponse({ description: MSG_FORBIDDEN })
    async put(
        @Body() werkstattDTO: WerkstattDtoOhneRef,
        @Param('id') id: number,
        @Headers('If-Match') version: string | undefined,
        @Res() res: Response,
    ): Promise<Response> {
        this.#logger.debug(
            'put: id=%s, werkstattDTO=%o, version=%s',
            id,
            werkstattDTO,
            version,
        );
        if (version === undefined) {
            const msg = 'Header "If-Match" fehlt';
            this.#logger.debug('put: msg=%s', msg);
            return res
                .status(HttpStatus.PRECONDITION_FAILED)
                .set('Content-Type', 'application/json')
                .send(msg);
        }

        const werkstatt = this.#werkstattDtoOhneRefToWerkstatt(werkstattDTO);
        const neueVersion = await this.#service.update({
            id,
            werkstatt,
            version,
        });
        this.#logger.debug('put: version=%s', neueVersion);
        return res.header('ETag', `"${neueVersion}"`).send();
    }

    @Delete(':id')
    @Roles({ roles: ['admin'] })
    @HttpCode(HttpStatus.NO_CONTENT)
    @ApiOperation({ summary: 'Die Werkstatt mit der ID löschen ' })
    @ApiNoContentResponse({
        description: 'Die Werkstatt wurde gelöscht oder war nicht vorhanden',
    })
    @ApiForbiddenResponse({ description: MSG_FORBIDDEN })
    async delete(@Param('id') id: number) {
        this.#logger.debug('delete: id=%s', id);
        await this.#service.delete(id);
    }

    #werkstattDtoToWerkstatt(werkstattDTO: WerkstattDTO): Werkstatt {
        const adresseDTO = werkstattDTO.adresse;
        const adresse: Adresse = {
            id: undefined,
            postleitzahl: adresseDTO.postleitzahl,
            strasse: adresseDTO.strasse,
            hausnummer: adresseDTO.hausnummer,
            werkstatt: undefined,
        };
        const angestellte = werkstattDTO.angestellte?.map((angestellterDTO) => {
            const angestellter: Angestellter = {
                id: undefined,
                name: angestellterDTO.name,
                position: angestellterDTO.position,
                werkstatt: undefined,
            };
            return angestellter;
        });
        const werkstatt = {
            id: undefined,
            version: undefined,
            werkstattname: werkstattDTO.werkstattname,
            rating: werkstattDTO.rating,
            art: werkstattDTO.art,
            homepage: werkstattDTO.homepage,
            schlagwoerter: werkstattDTO.schlagwoerter,
            adresse,
            angestellte,
            erzeugt: new Date(),
            aktualisiert: new Date(),
        };

        werkstatt.adresse.werkstatt = werkstatt;
        werkstatt.angestellte?.forEach((angestellter) => {
            angestellter.werkstatt = werkstatt;
        });
        return werkstatt;
    }

    #werkstattDtoOhneRefToWerkstatt(
        werkstattDTO: WerkstattDtoOhneRef,
    ): Werkstatt {
        return {
            id: undefined,
            version: undefined,
            werkstattname: werkstattDTO.werkstattname,
            rating: werkstattDTO.rating,
            art: werkstattDTO.art,
            homepage: werkstattDTO.homepage,
            schlagwoerter: werkstattDTO.schlagwoerter,
            adresse: undefined,
            angestellte: undefined,
            erzeugt: undefined,
            aktualisiert: new Date(),
        };
    }
}
