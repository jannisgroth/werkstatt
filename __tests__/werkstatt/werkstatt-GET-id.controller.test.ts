import { afterAll, beforeAll, describe, expect, test } from '@jest/globals';
import { HttpStatus } from '@nestjs/common';
import axios, { type AxiosInstance, type AxiosResponse } from 'axios';
import { type WerkstattModel } from '../../src/werkstatt/controller/werkstatt-get.controller.js';
import {
    host,
    httpsAgent,
    port,
    shutdownServer,
    startServer,
} from '../testserver.js';

// Testdaten
const idVorhanden = '1';

// Test-Suite
describe('GET /rest/:id', () => {
    let client: AxiosInstance;

    // Testserver starten und dabei mit der DB verbinden
    beforeAll(async () => {
        await startServer();
        const baseURL = `https://${host}:${port}/rest`;
        client = axios.create({
            baseURL,
            httpsAgent,
            validateStatus: (status) => status < 500, // eslint-disable-line @typescript-eslint/no-magic-numbers
        });
    });

    afterAll(async () => {
        await shutdownServer();
    });

    test('Buch zu vorhandener ID', async () => {
        // given
        const url = `/${idVorhanden}`;

        // when
        const { status, headers, data }: AxiosResponse<WerkstattModel> =
            await client.get(url);

        // then
        expect(status).toBe(HttpStatus.OK);
        expect(headers['content-type']).toMatch(/json/iu);

        // eslint-disable-next-line no-underscore-dangle
        const selfLink = data._links.self.href;

        expect(selfLink).toMatchSnapshot();
    });
});
