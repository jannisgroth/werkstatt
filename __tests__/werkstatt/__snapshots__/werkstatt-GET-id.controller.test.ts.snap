// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /rest/:id Buch zu vorhandener ID 1`] = `"https://jannisgroth:3000/rest/1"`;
